File: GuideObjC/GuideObjC/Objects/EdgeDownloader.m
//
//  EdgeDownloader.m
//  GuideObjC
//
//  Created by LuzanovRoman on 22.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

#import "EdgeDownloader.h"
#import "EGF2.h"

@interface EdgeDownloader ()
@property (nonatomic, retain) NSString *edge;
@property (nonatomic, retain) NSString *source;
@property (nonatomic, retain) NSArray *expand;
@property (nonatomic, retain) NSMutableArray *subscribedObjectIds;
@property (nonatomic, assign) BOOL downloading;
@end

@implementation EdgeDownloader

- (BOOL)noAnyData {
    return self.totalCount == -1;
}

- (BOOL)isDownloading {
    return _downloading;
}

- (id)initWithSource:(NSString *)source edge:(NSString *)edge {
    return [self initWithSource:source edge:edge expand:@[]];
}

- (id)initWithSource:(NSString *)source edge:(NSString *)edge expand:(NSArray *)expand {
    self = [self init];
    
    if (self) {
        self.expand = expand;
        self.source = source;
        self.edge = edge;
        self.pageCount = 25;
        self.subscribedObjectIds = [NSMutableArray array];
        [[Graph shared] addObserver:self selector:@selector(edgeCreated:) name:EGF2NotificationEdgeCreated forSource:source andEdge:edge];
        [[Graph shared] addObserver:self selector:@selector(edgeRemoved:) name:EGF2NotificationEdgeRemoved forSource:source andEdge:edge];
    }
    return self;
}

- (void)dealloc {
    [[Graph shared] removeObserver:self];
}

// MARK:- Private methods
- (void)edgeCreated:(NSNotification *)notification {
    NSString * objectId = notification.userInfo[EGF2EdgeObjectIdInfoKey];
    
    if (objectId) {
        __weak EdgeDownloader *weakSelf = self;
        
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.1 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
            if (weakSelf) {
                [[Graph shared] addObserver:weakSelf selector:@selector(objectUpdated:) name:EGF2NotificationObjectUpdated forSource:objectId];
                [[Graph shared] objectWithId:objectId expand:weakSelf.expandValues completion:^(NSObject * object, NSError * error) {
                    if (object) {
                        [weakSelf insertObject:object atIndex:0];
                    }
                }];
            }
        });
    }
}

- (void)edgeRemoved:(NSNotification *)notification {
    NSString * objectId = notification.userInfo[EGF2EdgeObjectIdInfoKey];
    
    if (objectId) {
        NSInteger index = [self indexOfObjectWithId:objectId];
        [self deleteObjectAtIndex:index];
        [[Graph shared] removeObserver:self name:EGF2NotificationObjectUpdated fromSource:objectId];
        [_subscribedObjectIds removeObject:objectId];
    }
}

- (void)objectUpdated:(NSNotification *)notification {
    NSString * objectId = notification.userInfo[EGF2ObjectIdInfoKey];
    
    if (objectId) {
        [[Graph shared] objectWithId:objectId expand:[self expandValues] completion:^(NSObject * object, NSError * error) {
            if (object) {
                [self replaceObject:object];
            }
        }];
    }
}

- (void)updateSubscriptionsForObjects:(NSArray *)objects  {
    NSMutableArray *newSubscribedObjectIds = [NSMutableArray array];
    
    for (NSObject *object in objects) {
        NSString *objectId =  [object valueForKey:@"id"];
        
        if (![_subscribedObjectIds containsObject:objectId]) {
            [newSubscribedObjectIds addObject:objectId];
        }
    }
    if (newSubscribedObjectIds.count > 0) {
        [[Graph shared] addObserver:self selector:@selector(objectUpdated:) name:EGF2NotificationObjectUpdated forSources:newSubscribedObjectIds];
    }
}

- (void)setObjects:(NSArray *)objects totalCount:(NSInteger)count {
    if (objects) {
        [self.graphObjects removeAllObjects];
        [self.tableView reloadData];
        [self addObjects:objects totalCount:count];
    }
}

- (void)addObjects:(NSArray *)objects totalCount:(NSInteger)count {
    if (objects) {
        BOOL addAnimated = self.graphObjects.count > 0;
        NSInteger start = self.graphObjects.count;
        NSInteger end = start + objects.count;
        NSMutableArray *indexPaths = [NSMutableArray array];
        
        for (NSInteger i = start; i < end; i++) {
            [indexPaths addObject:[NSIndexPath indexPathForRow:i inSection:0]];
        }
        self.totalCount = count;
        [self.graphObjects addObjectsFromArray:objects];
        
        if (addAnimated) {
            [self.tableView beginUpdates];
            [self.tableView insertRowsAtIndexPaths:indexPaths withRowAnimation:UITableViewRowAnimationNone];
            [self.tableView reloadRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:0 inSection:1]] withRowAnimation:UITableViewRowAnimationNone];
            [self.tableView endUpdates];
        }
        else {
            [self.tableView reloadData];
        }
        [self updateSubscriptionsForObjects:objects];
    }
}

- (void)insertObject:(NSObject *)object atIndex:(NSInteger)index {
    if (object) {
        // Check if object is already in the list
        if ([self indexOfObject:object] != NSNotFound) {
            return;
        }
        self.totalCount += 1;
        [self.graphObjects insertObject:object atIndex:index];
        
        [self.tableView beginUpdates];
        [self.tableView insertRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:index inSection:0]] withRowAnimation:UITableViewRowAnimationNone];
        [self.tableView endUpdates];
    }
}

- (void)deleteObjectAtIndex:(NSInteger)index {
    if (self.graphObjects.count == 0 || self.graphObjects.count <= index) {
        return;
    }
    self.totalCount -= 1;
    [self.graphObjects removeObjectAtIndex:index];
    
    [self.tableView beginUpdates];
    [self.tableView deleteRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:index inSection:0]] withRowAnimation:UITableViewRowAnimationNone];
    [self.tableView endUpdates];
}

- (void)deleteObject:(NSObject *)object {
    NSInteger index = [self.graphObjects indexOfObject:object];
    
    if (index != NSNotFound) {
        [self deleteObjectAtIndex:index];
    }
}

// MARK:- Override
- (NSArray *)expandValues {
    return _expand;
}

- (void)refreshList {
    if (_downloading) {
        return;
    }
    _downloading = true;
    [[Graph shared] refreshObjectsForSource:_source edge:_edge after:nil expand:_expand count:_pageCount completion:^(NSArray * objects, NSInteger count, NSError * error) {
        _downloading = false;
        
        [self setObjects:objects totalCount:count];
        [self.tableView.refreshControl endRefreshing];
    }];
}

- (void)getNextPage {
    if (_downloading) {
        return;
    }
    _downloading = true;
    [[Graph shared] objectsForSource:_source edge:_edge after:[self.last valueForKey:@"id"] expand:_expand count:_pageCount completion:^(NSArray * objects, NSInteger count, NSError * error) {
        _downloading = false;
        
        [self addObjects:objects totalCount:count];
    }];
}
@end

File: GuideSwift/GuideSwift/Objects/EdgeDownloader.swift
//
//  EdgeDownloader.swift
//  GuideSwift
//
//  Created by LuzanovRoman on 22.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

import UIKit
import EGF2

class EdgeDownloader<T: NSObject>: BaseDownloader<T> {

    fileprivate var subscribedObjectIds = [String]()
    fileprivate var downloading = false
    fileprivate var expand: [String]
    fileprivate var source: String
    fileprivate var edge: String

    var pageCount = 25

    var isDownloading: Bool {
        return downloading
    }

    var noAnyData: Bool {
        return totalCount == -1
    }

    init(withSource source: String, edge: String, expand: [String]? = nil) {
        self.expand = expand ?? []
        self.source = source
        self.edge = edge
        super.init()
        Graph.addObserver(self, selector: #selector(edgeCreated(notification:)), name: .EGF2EdgeCreated, forSource: source, andEdge: edge)
        Graph.addObserver(self, selector: #selector(edgeRemoved(notification:)), name: .EGF2EdgeRemoved, forSource: source, andEdge: edge)
    }
    
    deinit {
        Graph.removeObserver(self)
    }

    // MARK: - Private methods
    func edgeCreated(notification: NSNotification) {
        guard let objectId = notification.userInfo?[EGF2EdgeObjectIdInfoKey] as? String else { return }

        DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + .milliseconds(100)) { [weak self] in
            guard let strongSelf = self else { return }
            Graph.addObserver(strongSelf, selector: #selector(strongSelf.objectUpdated(notification:)), name: .EGF2ObjectUpdated, forSource: objectId)
            Graph.object(withId: objectId, expand: strongSelf.expand) { (object, _) in
                guard let graphObject = object else { return }
                strongSelf.insert(object: graphObject as? T, at: 0)
            }
        }
    }

    func edgeRemoved(notification: NSNotification) {
        guard let objectId = notification.userInfo?[EGF2EdgeObjectIdInfoKey] as? String else { return }
        guard let index = indexOfObject(withId: objectId) else { return }
        delete(at: index)
        Graph.removeObserver(self, name: .EGF2ObjectUpdated, fromSource: objectId)
        subscribedObjectIds.remove(objectId)
    }
    
    func objectUpdated(notification: NSNotification) {
        guard let objectId = notification.userInfo?[EGF2ObjectIdInfoKey] as? String else { return }

        Graph.object(withId: objectId, expand: expandValues) { (object, _) in
            guard let updatedObject = object as? T else { return }
            self.replace(object: updatedObject)
        }
    }

    fileprivate func set(objects: [T]?, totalCount count: Int) {
        guard let theObjects = objects else { return }
        graphObjects.removeAll()

        if let tv = tableView {
            tv.reloadData()
        }
        add(objects: theObjects, totalCount: count)
    }

    func add(objects: [T]?, totalCount count: Int) {
        guard let theObjects = objects else { return }

        let addAnimated = graphObjects.count > 0
        let start = graphObjects.count
        let end = start + theObjects.count
        var indexPaths = [IndexPath]()

        for i in start..<end {
            indexPaths.append(IndexPath(row: i, section: 0))
        }
        totalCount = count
        graphObjects.append(contentsOf: theObjects)

        guard let tv = tableView else { return }

        if addAnimated {
            tv.beginUpdates()
            tv.insertRows(at: indexPaths, with: .none)
            tv.reloadRows(at: [IndexPath(row: 0, section: 1)], with: .none)
            tv.endUpdates()
        } else {
            tv.reloadData()
        }
        updateSubscriptions(forObjects: theObjects)
    }
    
    func updateSubscriptions(forObjects objects: [T]) {
        var newSubscribedObjectIds = [String]()
        
        for object in objects {
            guard let objectId = object.value(forKey: "id") as? String else { return }
            if subscribedObjectIds.contains(objectId) { return }
            newSubscribedObjectIds.append(objectId)
        }
        if newSubscribedObjectIds.count > 0 {
            Graph.addObserver(self, selector: #selector(objectUpdated(notification:)), name: .EGF2ObjectUpdated, forSources: newSubscribedObjectIds)
        }
    }

    func insert(object: T?, at index: Int) {
        guard let theObject = object, let objectId = theObject.value(forKey: "id") as? String else { return }

        // Check if object is already in the list
        if let _ = indexOfObject(withId: objectId) {
            return
        }
        totalCount += 1
        graphObjects.insert(theObject, at: index)

        guard let tv = tableView else { return }
        tv.beginUpdates()
        tv.insertRows(at: [IndexPath(row: index, section: 0)], with: .none)
        tv.endUpdates()
    }

    func delete(at index: Int) {
        if graphObjects.count == 0 || graphObjects.count <= index { return }
        totalCount -= 1
        graphObjects.remove(at: index)
        
        guard let tv = tableView else { return }
        tv.beginUpdates()
        tv.deleteRows(at: [IndexPath(row: index, section: 0)], with: .none)
        tv.endUpdates()
    }
    
    func delete(object: T?) {
        guard let theObject = object, let index = graphObjects.index(of: theObject) else { return }
        delete(at: index)
    }

    // MARK: - Override
    override var expandValues: [String] {
        return expand
    }

    override func refreshList() {
        if downloading { return }

        downloading = true
        Graph.refreshObjects(forSource: source, edge: edge, after: nil, expand: expand, count: pageCount) { (objects, count, _) in
            self.downloading = false
            self.tableView?.refreshControl?.endRefreshing()
            self.set(objects: objects as? [T], totalCount: count)
        }
    }

    override func getNextPage() {
        if downloading { return }

        downloading = true
        Graph.objects(forSource: source, edge: edge, after: self.last?.value(forKey: "id") as? String, expand: expand, count: pageCount) { (objects, count, _) in
            self.downloading = false
            self.add(objects: objects as? [T], totalCount: count)
        }
    }
}
